KeyError with unsupported Slurm version

Error reported by @Talavig in #315, with unsupported Slurm version:

```
Thread-38 (process_request_thread): [INFO] 127.0.0.1 - - [16/Jul/2024 13:28:26] "GET /v3.1.0/stats HTTP/1.1" 500 -
Jul 16 13:28:26 standalone slurm-web-agent[2310]: KeyError: 'warnings'
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     if len(result["warnings"]):
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/slurmweb/views/agent.py", line 64, in slurmrest
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     items = func(*args)
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/slurmweb/views/agent.py", line 76, in filter_fields
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     return func(*args)
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/slurmweb/views/agent.py", line 88, in _cached_data
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     return _cached_data(
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/slurmweb/views/agent.py", line 101, in _cached_jobs
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     for job in _cached_jobs():
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/slurmweb/views/agent.py", line 236, in stats
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     return view(*args, **kwargs)
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/rfl/web/tokens.py", line 93, in wrapped
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/flask/app.py", line 1499, in dispatch_request
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     rv = self.dispatch_request()
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/flask/app.py", line 1513, in full_dispatch_request
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     rv = self.handle_user_exception(e)
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/flask/app.py", line 1515, in full_dispatch_request
Jul 16 13:28:26 standalone slurm-web-agent[2310]:     response = self.full_dispatch_request()
Jul 16 13:28:26 standalone slurm-web-agent[2310]:   File "/usr/lib/python3/dist-packages/flask/app.py", line 2070, in wsgi_app
Jul 16 13:28:26 standalone slurm-web-agent[2310]: Traceback (most recent call last):
Jul 16 13:28:26 standalone slurm-web-agent[2310]: Thread-38 (process_request_thread): [ERROR] Exception on /v3.1.0/stats [GET]
```

Unsupported Slurm version error should be reported more clearly to users.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -31,6 +31,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## Changed
 - frontend: Add intermediate cluster list width to 80% on large screens, before
   going down to 60% on even larger screens.
+- agent: Check Slurm version returned from `slurmrestd` against hard-coded
+  minimal version and log error if not greater or equal.
 - pkgs: Add requirement on RFL.core and RFL.authentication >= 1.0.3.
 - docs: Update configuration reference documentation.
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -44,6 +44,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
     interpreted by frontend and emit clear error message in logs (#321).
   - Detect responses from slurmrestd not formatted in JSON, translated into JSON
     error for frontend and emit clear error message in logs (#333).
+  - Detect absence of _warnings_ key in `slurmrestd` responses and emit warning
+    log instead of crashing (#316).
 - genjwt: fix portability to Python < 3.8 in debug message.
 - ldap-check: fix usage of `user_name_attribute` configuration parameter (#340).
 - frontend:

--- a/slurmweb/views/agent.py
+++ b/slurmweb/views/agent.py
@@ -15,6 +15,9 @@ from ..version import get_version
 from ..errors import SlurmwebCacheError, SlurmwebRestdError
 from . import SlurmrestdUnixAdapter
 
+# Tuple used for comparaison with Slurm version retrieved from slurmrestd and
+# check for minimal supported version.
+MINIMAL_SLURM_VERSION = (23, 2, 0)
 
 logger = logging.getLogger(__name__)
 
@@ -137,7 +140,7 @@ def _cached_data(cache_key: str, expiration: int, func: Callable, *args: List[An
 def _get_version():
     return slurmrest(f"/slurm/v{current_app.settings.slurmrestd.version}/ping", "meta")[
         "Slurm"
-    ]["release"]
+    ]
 
 
 def _cached_version():
@@ -284,6 +287,22 @@ def _cached_accounts():
 def stats():
     total = 0
     running = 0
+
+    version = _cached_version()
+    logger.info("Retrieved version %s", version)
+    # Check Slurm version is supported or fail with HTTP/500
+    if (
+        not (
+            version["version"]["major"],
+            version["version"]["minor"],
+            version["version"]["micro"],
+        )
+        >= MINIMAL_SLURM_VERSION
+    ):
+        error = f"Unsupported Slurm version {version['release']}"
+        logger.error(error)
+        abort(500, error)
+
     for job in _cached_jobs():
         total += 1
         if "RUNNING" in job["job_state"]:
@@ -296,7 +315,7 @@ def stats():
         cores += node["cpus"]
     return jsonify(
         {
-            "version": _cached_version(),
+            "version": version["release"],
             "resources": {"nodes": nodes, "cores": cores},
             "jobs": {"running": running, "total": total},
         }

--- a/slurmweb/views/agent.py
+++ b/slurmweb/views/agent.py
@@ -92,7 +92,13 @@ def slurmrest(query, key, handle_errors=True):
                 error["description"],
                 error["source"],
             )
-    if len(result["warnings"]):
+    if "warnings" not in result:
+        logger.error(
+            "Unable to extract warnings from slurmrestd response to %s, unsupported "
+            "Slurm version?",
+            query,
+        )
+    elif len(result["warnings"]):
         logger.warning("slurmrestd query %s warnings: %s", query, result["warnings"])
     return result[key]
 

